### PR TITLE
FinalizeReplicaUpdate

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
Closes https://github.com/filecoin-project/lotus/issues/8012

Depends on https://github.com/filecoin-project/specs-storage/pull/22

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

File: extern/sector-storage/ffiwrapper/sealer_cgo.go [hunk 1]
```diff
@@ -843,6 +843,82 @@ func (sb *Sealer) FinalizeSector(ctx context.Context, sector storage.SectorRef,
...
+
+	paths, done, err := sb.sectors.AcquireSector(ctx, sector, storiface.FTCache, 0, storiface.PathStorage)
+	if err != nil {
+		return xerrors.Errorf("acquiring sector cache path: %w", err)
+	}
+	defer done()
+
+	return ffi.ClearCache(uint64(ssize), paths.Cache)
+
+	// TODO: ^ above but for snapdeals
```
Comment (magik6k) [inline extern/sector-storage/ffiwrapper/sealer_cgo.go:919]: Main todo; not sure if passing UpdateCache into ClearCache is/does the correct thing

File: extern/sector-storage/ffiwrapper/sealer_cgo.go [hunk 2]
```diff
@@ -843,6 +843,82 @@ func (sb *Sealer) FinalizeSector(ctx context.Context, sector storage.SectorRef,
...
 }
 
+func (sb *Sealer) FinalizeReplicaUpdate(ctx context.Context, sector storage.SectorRef, keepUnsealed []storage.Range) error {
+	ssize, err := sector.ProofType.SectorSize()
+	if err != nil {
+		return err
+	}
+	maxPieceSize := abi.PaddedPieceSize(ssize)
+
+	if len(keepUnsealed) > 0 { // TODO dedupe with the above
```
Comment (magik6k) [inline extern/sector-storage/ffiwrapper/sealer_cgo.go:853]: Should be done before merging

**Review by ZenGround0 — APPROVED** [review]
LGTM reminder to update specs-storage dependency to master

One thing this reminds me of is that we should probably track work for calling manager ReleaseSectorKey during the finalize step based on a config param.  I think right now we end up carrying original Replica, UpdatedReplica and UnsealedData by default.  And I don't think there's a good way to configure dropping the sector key.

**Comment by remakeZK** [comment]
Should old sealed files keep in the path until `ChainFanilty`?Since winning post check the sector state in the past, clear old sealed immediately may cause some unexcept problem

**Comment by magik6k** [comment]
> Should old sealed files keep in the path until ChainFanilty?Since winning post check the sector state in the past, clear old sealed immediately may cause some unexcept problem

They are kept, this PR doesn't handle cleanup of those, this will come as a follow-up

**Comment by magik6k** [comment]
Or maybe that's easy enough to do now

**Comment by ZenGround0** [comment]
Yeah thanks @remakeZK, I was forgetting about winning post.  We'll need a separate state to do sector key cleanup, or postpone finalize by 2 finalities and do it there.

File: extern/storage-sealing/states_replica_update.go [hunk 3]
```diff
@@ -203,9 +206,87 @@ func (m *Sealing) handleReplicaUpdateWait(ctx statemachine.Context, sector Secto
...
+				break
+			}
+
+			wctx, cancel := context.WithTimeout(ctx.Context(), delay)
+			<-wctx.Done()
+			cancel()
+
+			// increasing backoff; can't just calculate the correct time because integration tests do funny things with time
+			delay = delay * 10 / 3
+			if delay > 5*time.Minute {
```
Comment (ZenGround0) [inline extern/storage-sealing/states_replica_update.go:254]: What's the benefit of capping at 5 minutes?  Isn't the winning post lookup about a day long?

**Review by ZenGround0 — APPROVED** [review]

File: extern/storage-sealing/states_replica_update.go (hunk 3 above)
Comment (magik6k) [reply extern/storage-sealing/states_replica_update.go:254]: Hrm, there is a better way

**Comment by jennijuju** [comment]
Need better PR title

**Pull request merged**